apps/x509.c

```diff
@@ -730,7 +730,7 @@ int x509_main(int argc, char **argv)
         }
         if ((x = X509_new_ex(app_get0_libctx(), app_get0_propq())) == NULL)
             goto end;
-        if (sno == NULL) {
+        if (CAfile == NULL && sno == NULL) {
```

Comment:
Should this be CAserial == NULL instead?

Reply:
Or actually `CAserial == NULL && CA_createserial == 0 && sno == NULL`

Reply:
`CAfile == NULL && sno == NULL` is correct and sufficient because the guard for the later call
```
            sno = x509_load_serial(CAfile, CAserial, CA_createserial);
```
is `CAfile != NULL`.

Using `CAserial == NULL`  (and `CA_createserial == 0`) would be wrong here because `x509_load_serial()` does whatever is needed also in case any of these two conditions hold,
and in case `CAserial == NULL` (and possibly in addition `CA_createserial == 0`) the need for a fallback random number depends on the existence of the fallback serial number file derived from `CAfile`.

Reply:
So perhaps `CAfile == NULL` is right condition here but then further fixes are needed in x509_load_serial or load_serial.

Reply:
Why? 
I had a look again at their implementations.

If `CAfile != NULL` then `x509_load_serial()` is called.
The `serialfile` to use is implicit in case `CAserial` is not provided.
* If this file does not exist and `CA_createserial != 0`, a random number is used as expected and documented, both for `x509` and `ca` (which is the other app using `load_serial()`).
* If this file does not exist and `CA_createserial == 0`, a random number is not used, again as documented for both apps. Note that your above quotation from `openssl-x509.pod.in` only pertains to the case that the `-CAcreateserial` option is specified.

Reply:
I don't have a hand in the codebase, but I use openssl for something every week.

Given that this is trying to fix some accidental breakage from a different PR (https://github.com/openssl/openssl/pull/13711#issuecomment-1131985954) wouldn't it be safer to take the defensive route and just add

> actually `CAserial == NULL && CA_createserial == 0 && sno == NULL`

instead of trying to optimize and minimize all possible code paths mentally?

Reply:
> The documentation is really misleading here. Because IMO the semantics should really be to use a random number if the serial number file does not exist and `CA_createserial == 0`. 

Although I agree that using a random serial number generally is good as a fallback,
I cannot really agree here.
If the user specifies a `-CAserial` file and does not give `-CAcreateserial` and the specified file does not exist, this clearly is an error (likely the user made a mistake with the file name), 
and IMO it should be treated as such.

Reply:
We are talking about the `CAserial == NULL` case here!

Reply:
First of all, I noticed that my title (but not the OP) of this PR was wrong, and corrected this.

> We are talking about the `CAserial == NULL` case here!

According to your clarification, it turns out that in our discussion here I was having a subtle misunderstanding, since before you were just writing:

> The documentation is really misleading here. Because IMO the semantics should really be to use a random number if the serial number file does not exist and `CA_createserial == 0`. 

and because "the serial number file does not exist" is different from `CAserial == NULL`.

`CAserial == NULL` means that the user did not specify the `-CAserial` option,
while this PR (so far) was just about correcting the case `CAserial != NULL`.

In case `CAserial == NULL` (and `-CA` is given but `-set_serial` is not given),
`x509_load_serial()` takes a default serial number file name derived from `CAfile`.
I now see the case you aimed at improving the situation that in case this **default** serial number file does not exist and `CA_createserial == 0`, an error is thrown. 
I agree that in this case it is better to generate and use a random serial number instead.
I've improved the documentation accordingly and implemented this.

Doing so, I found that that the part of the documentation saying
> it will contain the serial number "02" and the certificate being signed will
> have the 1 as its serial number.

is not (or no more) valid - I replaced it by
"a random number is generated and stored; this is the recommended practice."

On this occasion, I also improved the coding style of `load_serial()` in  `apps.c`
and fixed the error message and a memory leak in case `rand_serial()` failed.